### PR TITLE
Use lazy cancellation in RemoteEndPoint.InvokeAsync<T>

### PR DESCRIPTION
**Pull request by sharwell**
This change ensures that SolutionAssetStorage does not release a pinned
data scope while it is still in use.

Fixes [AB#1246701](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1246701)

(cherry picked from commit 745ddafff5d52e8a907e46c3c55ea8906e095270)

### CODE REVIEW

File: src/Workspaces/Remote/Core/RemoteEndPoint.cs [hunk 1]
```diff
@@ -196,16 +196,22 @@ public async Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object?> ar
...
+                    // run user task with direct stream
+                    return await dataReader(stream, linkedCancellationSource.Token).ConfigureAwait(false);
+                }, linkedCancellationSource.Token);
 
-                // wait task to finish
-                await task.ConfigureAwait(false);
+                // Wait for all tasks to finish. If we return while one of the tasks is still in flight, the task would
+                // operate as a fire-and-forget operation where the caller might release state objects still in use by
+                // the task.
+                await Task.WhenAll(task, task2).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/RemoteEndPoint.cs:211]: wasn't this change already done?  this looks super familiar.

Comment (sharwell) [reply src/Workspaces/Remote/Core/RemoteEndPoint.cs:211]: This is a back-port of #49164 to release/dev16.8

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/Core/RemoteEndPoint.cs:211]: ah ok :)

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by jinujoseph** [comment]
Approved to be in 16.8.2

**Pull request merged**